File: mg_slam/scripts/slam_gnss_2d/input/time_series.py
from __future__ import annotations

import bisect
from typing import Optional, Sequence, TypeVar

from slam_gnss_2d.core.data_types import GnssData, OdomData
from slam_gnss_2d.core.geometry import angle_diff

T = TypeVar('T')


def nearest_by_timestamp(
    items: Sequence[T],
    timestamps: Sequence[float],
    timestamp: float,
    max_dt: Optional[float] = None,
) -> Optional[T]:
    """時刻に最も近い要素を返す。max_dt が指定されている場合は許容時間差を超えたら None を返す。"""
    if not items:
        return None
    idx = bisect.bisect_left(timestamps, timestamp)
    if idx == 0:
        best = items[0]
        best_dt = abs(timestamp - timestamps[0])
    elif idx >= len(items):
        best = items[-1]
        best_dt = abs(timestamp - timestamps[-1])
    else:
        dt_prev = abs(timestamp - timestamps[idx - 1])
        dt_next = abs(timestamp - timestamps[idx])
        if dt_prev <= dt_next:
            best = items[idx - 1]
            best_dt = dt_prev
        else:
            best = items[idx]
            best_dt = dt_next
    if max_dt is not None and best_dt > max_dt:
        return None
    return best


def interpolate_odom(
    odom_list: Sequence[OdomData],
    timestamps: Sequence[float],
    timestamp: float,
) -> Optional[OdomData]:
    """OdomData を指定時刻へ線形補間して返す。"""
    if not odom_list:
        return None
    idx = bisect.bisect_left(timestamps, timestamp)
    if idx == 0:
        return odom_list[0]
    if idx >= len(odom_list):
        return odom_list[-1]
    prev = odom_list[idx - 1]
    next_ = odom_list[idx]
    t_span = next_.timestamp - prev.timestamp
    if t_span < 1e-9:
        return prev
    alpha = (timestamp - prev.timestamp) / t_span
    return OdomData(
        timestamp=timestamp,
        x=prev.x + alpha * (next_.x - prev.x),
        y=prev.y + alpha * (next_.y - prev.y),
        yaw=prev.yaw + alpha * angle_diff(next_.yaw, prev.yaw),
    )


def interpolate_gnss(
    gnss_list: Sequence[GnssData],
    timestamps: Sequence[float],
    timestamp: float,
    max_dt: Optional[float] = None,
) -> Optional[GnssData]:
    """GnssData を指定時刻へ線形補間して返す。max_dt が指定されている場合は許容時間差を超えたら None を返す。"""
    if not gnss_list:
        return None
    idx = bisect.bisect_left(timestamps, timestamp)
    if idx == 0:
        if max_dt is not None and abs(timestamp - timestamps[0]) > max_dt:
            return None
        return gnss_list[0]
    if idx >= len(gnss_list):
        if max_dt is not None and abs(timestamp - timestamps[-1]) > max_dt:
            return None
        return gnss_list[-1]
    prev = gnss_list[idx - 1]
    next_ = gnss_list[idx]
    t_span = next_.timestamp - prev.timestamp
    if t_span < 1e-9:
        return prev
    if max_dt is not None and (t_span > max_dt or abs(timestamp - prev.timestamp) > max_dt or abs(timestamp - next_.timestamp) > max_dt):
        return None
    alpha = (timestamp - prev.timestamp) / t_span
    return GnssData(
        timestamp=timestamp,
        x=prev.x + alpha * (next_.x - prev.x),
        y=prev.y + alpha * (next_.y - prev.y),
        covariance=prev.covariance + alpha * (next_.covariance - prev.covariance),
        fix_status=prev.fix_status,
        latitude=prev.latitude + alpha * (next_.latitude - prev.latitude),
        longitude=prev.longitude + alpha * (next_.longitude - prev.longitude),
    )

File: slam_gnss_2d/slam_gnss_2d/src/input/time_series.cpp
#include "slam_gnss_2d/input/time_series.hpp"

namespace slam_gnss_2d {
namespace input {

std::optional<core::OdomData> interpolate_odom(
    const std::vector<core::OdomData>& odom_list,
    const std::vector<double>& timestamps,
    double timestamp) {
  if (odom_list.empty() || timestamps.empty() || odom_list.size() != timestamps.size()) {
    return std::nullopt;
  }
  auto it = std::lower_bound(timestamps.begin(), timestamps.end(), timestamp);
  size_t idx = std::distance(timestamps.begin(), it);

  if (idx == 0) {
    return odom_list[0];
  }
  if (idx >= odom_list.size()) {
    return odom_list.back();
  }

  const auto& prev = odom_list[idx - 1];
  const auto& next = odom_list[idx];
  double t_span = next.timestamp - prev.timestamp;
  if (t_span < 1e-9) {
    return prev;
  }

  double alpha = (timestamp - prev.timestamp) / t_span;
  double interp_x = prev.x + alpha * (next.x - prev.x);
  double interp_y = prev.y + alpha * (next.y - prev.y);
  double interp_yaw = prev.yaw + alpha * core::angle_diff(next.yaw, prev.yaw);

  return core::OdomData{timestamp, interp_x, interp_y, interp_yaw};
}

std::optional<core::GnssData> interpolate_gnss(
    const std::vector<core::GnssData>& gnss_list,
    const std::vector<double>& timestamps,
    double timestamp,
    std::optional<double> max_dt) {
  if (gnss_list.empty() || timestamps.empty() || gnss_list.size() != timestamps.size()) {
    return std::nullopt;
  }
  auto it = std::lower_bound(timestamps.begin(), timestamps.end(), timestamp);
  size_t idx = std::distance(timestamps.begin(), it);

  if (idx == 0) {
    if (max_dt.has_value() && std::abs(timestamp - timestamps[0]) > *max_dt) {
      return std::nullopt;
    }
    return gnss_list[0];
  }
  if (idx >= gnss_list.size()) {
    if (max_dt.has_value() && std::abs(timestamp - timestamps.back()) > *max_dt) {
      return std::nullopt;
    }
    return gnss_list.back();
  }

  const auto& prev = gnss_list[idx - 1];
  const auto& next = gnss_list[idx];
  double t_span = next.timestamp - prev.timestamp;
  if (t_span < 1e-9) {
    return prev;
  }

  if (max_dt.has_value() &&
      (t_span > *max_dt ||
       std::abs(timestamp - prev.timestamp) > *max_dt ||
       std::abs(timestamp - next.timestamp) > *max_dt)) {
    return std::nullopt;
  }

  double alpha = (timestamp - prev.timestamp) / t_span;
  core::GnssData result;
  result.timestamp = timestamp;
  result.x = prev.x + alpha * (next.x - prev.x);
  result.y = prev.y + alpha * (next.y - prev.y);
  result.covariance = prev.covariance + alpha * (next.covariance - prev.covariance);
  result.fix_status = prev.fix_status;
  result.latitude = prev.latitude + alpha * (next.latitude - prev.latitude);
  result.longitude = prev.longitude + alpha * (next.longitude - prev.longitude);

  return result;
}

}  // namespace input
}  // namespace slam_gnss_2d
